File: internal/model/servers.go
package model

import (
	"encoding/hex"
	"fmt"
	"net"
	"net/url"
	"strings"
	"time"
	"unicode/utf8"
)

// ServerConnection is another NodeHoster server managed from this one, like
// a connection added with IIS Manager's "Connect to a Server": the web
// console of this server can switch to it and operate it through an
// authenticated proxy, and the Servers page shows its health.
//
// The connection authenticates with an API token created on the remote
// server (a dedicated one, limited to what this server's users should do
// there). What the token allows is enforced there; MinRole decides which
// of this server's users may use the connection at all.
type ServerConnection struct {
	ID   string `json:"id"`
	Name string `json:"name"`
	// URL is the remote web console's base URL, e.g. https://web02:8484
	// (a path prefix is kept when the console sits behind a reverse
	// proxy). Its API is URL + /api.
	URL   string `json:"url"`
	Token string `json:"token"` // secret: an API token created on the remote server
	// Fingerprint pins the remote certificate (SHA-256 of the leaf, 64
	// hex digits) instead of verifying it against the trusted roots, for
	// the self-signed certificates admin listeners often have. Empty: the
	// certificate must be valid for the host name.
	Fingerprint string `json:"fingerprint,omitempty"`
	// MinRole is the least local role allowed to use the connection:
	// admin (the default), operator or viewer. The remote server caps the
	// token at the caller's local role (see RoleLimitHeader); one too old
	// to do so can only be used by administrators.
	MinRole   Role      `json:"minRole"`
	CreatedAt time.Time `json:"createdAt"`
	UpdatedAt time.Time `json:"updatedAt"`
}

// RoleLimitHeader asks the server to narrow the caller's access to at most
// the given role for this request. It can only take rights away, so any
// client may send it; the proxy of a server connection sends the local
// user's role, so a local operator cannot act as an administrator on the
// remote server even with an administrator's token.
const RoleLimitHeader = "X-NodeHoster-Role-Limit"

// RoleLimitAppliedHeader is the answer to RoleLimitHeader: a server that
// applied the limit echoes the role in it. A server too old to know the
// limit ignores it and answers without, so the proxy of a connection
// relays a non-administrator's request only to a server that echoes it.
const RoleLimitAppliedHeader = "X-NodeHoster-Role-Limit-Applied"

// ServerHealth is what the last check of a connection found.
type ServerHealth struct {
	// Reachable: the server answered with the token accepted. False with
	// CheckedAt nil means the connection was not checked yet.
	Reachable bool       `json:"reachable"`
	Error     string     `json:"error,omitempty"`
	CheckedAt *time.Time `json:"checkedAt,omitempty"`
	LatencyMs int64      `json:"latencyMs"`
	// Since when the server has been in its current state (reachable or
	// not), as seen from here.
	Since *time.Time `json:"since,omitempty"`

	Version    string  `json:"version,omitempty"` // "" when the server does not say
	Commit     string  `json:"commit,omitempty"`
	Hostname   string  `json:"hostname,omitempty"`
	OS         string  `json:"os,omitempty"`
	CPUPercent float64 `json:"cpuPercent"`
	CPUCount   int     `json:"cpuCount"`
	MemTotal   uint64  `json:"memTotal"`
	MemUsed    uint64  `json:"memUsed"`

	// The sites the token can see, by state.
	Sites    int `json:"sites"`
	Running  int `json:"running"`
	Degraded int `json:"degraded"`
	Failed   int `json:"failed"`
	Stopped  int `json:"stopped"`

	// Who the token is on the remote server, and its role there
	// (RoleSites for a token limited to some sites).
	User string `json:"user,omitempty"`
	Role Role   `json:"role,omitempty"`
	// RoleLimits: the server applies RoleLimitHeader. Known when User is
	// set (the check read who the token is); a server without it can
	// only be used by this server's administrators.
	RoleLimits bool `json:"roleLimits"`
}

// ServerView is a connection as the API lists it: the token masked, with
// its health.
type ServerView struct {
	ServerConnection
	Health ServerHealth `json:"health"`
}

// ServerTest is the body of POST /api/servers/test: a connection being
// set up. With ID set and Token masked, the stored token is used.
type ServerTest struct {
	ID          string `json:"id,omitempty"`
	URL         string `json:"url"`
	Token       string `json:"token"`
	Fingerprint string `json:"fingerprint,omitempty"`
}

// ServerTestResult is what testing a connection found: the certificate
// the server presented (for trust on first use) and, when the TLS
// connection could be trusted, the server's answer to the token.
type ServerTestResult struct {
	Certificate *PeerCertificate `json:"certificate,omitempty"` // nil over plain HTTP
	// Trusted: the connection's TLS settings accept the certificate (it
	// matches the pinned fingerprint, or verifies against the trusted
	// roots when none is pinned). Nothing is sent with the token otherwise.
	Trusted bool         `json:"trusted"`
	Health  ServerHealth `json:"health"`
}

// PeerCertificate describes the certificate a server presented.
type PeerCertificate struct {
	Fingerprint string    `json:"fingerprint"` // SHA-256, uppercase hex
	Subject     string    `json:"subject"`
	Issuer      string    `json:"issuer"`
	DNSNames    []string  `json:"dnsNames"`
	NotBefore   time.Time `json:"notBefore"`
	NotAfter    time.Time `json:"notAfter"`
	// Verified: the chain verifies for the host name against the trusted
	// roots of the computer that connected; VerifyError says why not.
	Verified    bool   `json:"verified"`
	VerifyError string `json:"verifyError,omitempty"`
}

// ApplyDefaults normalizes a connection as entered.
func (s *ServerConnection) ApplyDefaults() {
	s.Name = strings.TrimSpace(s.Name)
	s.URL = strings.TrimSpace(s.URL)
	s.Token = strings.TrimSpace(s.Token)
	if u, err := NormalizeServerURL(s.URL); err == nil {
		s.URL = u
	}
	if fp, err := ParseFingerprint(s.Fingerprint); err == nil {
		s.Fingerprint = fp
	}
	if s.MinRole == "" {
		s.MinRole = RoleAdmin
	}
}

// Validate checks a normalized connection; others are the other
// connections, whose names must differ.
func (s *ServerConnection) Validate(others []ServerConnection) error {
	if s.Name == "" {
		return verr("name", "enter a name for the server")
	}
	if utf8.RuneCountInString(s.Name) > 64 {
		return verr("name", "at most 64 characters")
	}
	for _, o := range others {
		if o.ID != s.ID && strings.EqualFold(o.Name, s.Name) {
			return verr("name", "another server connection is named %q", o.Name)
		}
	}
	u, err := NormalizeServerURL(s.URL)
	if err != nil {
		return verr("url", "%v", err)
	}
	if _, err := ParseFingerprint(s.Fingerprint); err != nil {
		return verr("fingerprint", "%v", err)
	}
	if s.Fingerprint != "" && strings.HasPrefix(u, "http://") {
		return verr("fingerprint", "a certificate fingerprint needs an https:// URL")
	}
	if s.Token == "" {
		return verr("token", "enter an API token created on that server")
	}
	if !serverRole(s.MinRole) {
		return verr("minRole", "admin, operator or viewer")
	}
	return nil
}

// NormalizeServerURL checks the base URL of a NodeHoster web console and
// returns it without a trailing slash. HTTPS is required, except to this
// computer: the API token would otherwise cross the network in clear.
func NormalizeServerURL(raw string) (string, error) {
	raw = strings.TrimSpace(raw)
	if raw == "" {
		return "", fmt.Errorf("enter the server's web console URL, e.g. https://web02:8484")
	}
	if !strings.Contains(raw, "://") {
		raw = "https://" + raw
	}
	u, err := url.Parse(raw)
	if err != nil || u.Host == "" || u.Hostname() == "" {
		return "", fmt.Errorf("enter the server's web console URL, e.g. https://web02:8484")
	}
	if u.User != nil || u.RawQuery != "" || u.ForceQuery || u.Fragment != "" {
		return "", fmt.Errorf("the URL cannot have a user name, a query or a fragment")
	}
	switch strings.ToLower(u.Scheme) {
	case "https":
	case "http":
		if !IsLoopbackHost(u.Hostname()) {
			return "", fmt.Errorf("use https://: the API token must not cross the network in clear text")
		}
	default:
		return "", fmt.Errorf("the URL must start with https://")
	}
	if p := u.Port(); p != "" {
		var n int
		if _, err := fmt.Sscanf(p, "%d", &n); err != nil || n < 1 || n > 65535 {
			return "", fmt.Errorf("port %s is not valid", p)
		}
	}
	path := strings.TrimRight(u.EscapedPath(), "/")
	for _, seg := range strings.Split(path, "/") {
		if seg == "." || seg == ".." {
			return "", fmt.Errorf("the URL's path cannot contain . or ..")
		}
	}
	path = strings.TrimSuffix(path, "/api") // the API's own URL was entered
	return strings.ToLower(u.Scheme) + "://" + strings.ToLower(u.Host) + path, nil
}

// IsLoopbackHost reports whether a host name is this computer.
func IsLoopbackHost(h string) bool {
	if strings.EqualFold(h, "localhost") {
		return true
	}
	ip := net.ParseIP(strings.Trim(h, "[]"))
	return ip != nil && ip.IsLoopback()
}

// ParseFingerprint normalizes a SHA-256 certificate fingerprint as tools
// print it (AB:CD:…, ab cd …, SHA256:abcd…) to 64 uppercase hex digits.
// An empty string stays empty (no pinning).
func ParseFingerprint(s string) (string, error) {
	s = strings.TrimSpace(s)
	if s == "" {
		return "", nil
	}
	if i := strings.IndexByte(s, ':'); i > 0 && strings.EqualFold(strings.ReplaceAll(s[:i], "-", ""), "sha256") {
		s = s[i+1:]
	}
	s = strings.NewReplacer(":", "", " ", "", "-", "").Replace(s)
	b, err := hex.DecodeString(s)
	if err != nil || len(b) != 32 {
		return "", fmt.Errorf("a SHA-256 fingerprint is 64 hexadecimal digits (colons allowed)")
	}
	return strings.ToUpper(s), nil
}

// FormatFingerprint groups a fingerprint by bytes (AB:CD:…), for reading
// it out against the remote server's.
func FormatFingerprint(fp string) string {
	if len(fp)%2 != 0 {
		return fp
	}
	parts := make([]string, 0, len(fp)/2)
	for i := 0; i < len(fp); i += 2 {
		parts = append(parts, fp[i:i+2])
	}
	return strings.Join(parts, ":")
}

// RoleAtLeast reports whether role r is at least need (viewer < operator
// < admin). Site-scoped and unknown roles are below viewer.
func RoleAtLeast(r, need Role) bool {
	rank := func(r Role) int {
		switch r {
		case RoleViewer:
			return 1
		case RoleOperator:
			return 2
		case RoleAdmin:
			return 3
		}
		return 0
	}
	return rank(need) > 0 && rank(r) >= rank(need)
}

File: web/src/lib/servers.ts
// Multi-server management: which requests go to a connected server through
// this one's proxy, what the signed-in user may do there, and how a
// server's health reads. The server enforces all of it; this only decides
// where requests go and what the console shows.
import type { Access, Role, ServerConnection, ServerHealth } from '@/api/types';

/**
 * API paths that always stay on this server, whichever server the console
 * operates: the account (sign-in, password, two-factor, API tokens) and
 * the connections themselves (including their proxy).
 */
const LOCAL_PREFIXES = ['/api/auth/', '/api/tokens', '/api/servers'];

export function isLocalApiPath(path: string): boolean {
  if (!path.startsWith('/api/')) return true;
  return LOCAL_PREFIXES.some((p) => path === p || path.startsWith(p.endsWith('/') ? p : `${p}/`) || path.startsWith(`${p}?`));
}

/** The URL of an API path on a connected server (serverId), or on this one (null). */
export function routeApiPath(path: string, serverId: string | null): string {
  if (!serverId || isLocalApiPath(path)) return path;
  return `/api/servers/${encodeURIComponent(serverId)}/proxy/${path.slice('/api/'.length)}`;
}

/** Whether a URL goes through a connection's proxy. */
export function isProxiedUrl(url: string): boolean {
  return /^\/api\/servers\/[^/]+\/proxy\//.test(url);
}

const RANK: Record<string, number> = { viewer: 1, operator: 2, admin: 3 };

/**
 * What the console may offer on a connected server: the token's access
 * there, never above the local user's role. A remote server of this
 * version applies the same cap itself; for an older one, the proxy still
 * keeps local viewers read-only.
 */
export function capAccess(remote: Access | undefined, localRole: Role | undefined): Access | undefined {
  if (!remote || !localRole) return undefined;
  const limit = RANK[localRole] ?? 0;
  if (limit === 0) return { role: 'sites', sites: [] };
  const cap = <R extends string>(r: R): R => ((RANK[r] ?? 0) > limit ? (localRole as unknown as R) : r);
  if (remote.role === 'sites') {
    return { role: 'sites', sites: (remote.sites ?? []).map((g) => ({ ...g, role: (RANK[g.role] ?? 0) > limit ? 'viewer' : g.role })) };
  }
  return { role: cap(remote.role) };
}

/** Whether a local role may use a connection (its minimum role). */
export function mayUseServer(localRole: Role | undefined, s: Pick<ServerConnection, 'minRole'>): boolean {
  if (!localRole || localRole === 'sites') return false;
  return (RANK[localRole] ?? 0) >= (RANK[s.minRole] ?? 3);
}

export type HealthState = 'online' | 'warning' | 'offline' | 'unknown';

/** A server's state for the overview: offline, online, or online with failed sites. */
export function healthState(h: ServerHealth | undefined): HealthState {
  if (!h || !h.checkedAt) return 'unknown';
  if (!h.reachable) return 'offline';
  if (h.failed > 0 || h.degraded > 0) return 'warning';
  return 'online';
}

export function healthLabel(h: ServerHealth | undefined): string {
  switch (healthState(h)) {
    case 'unknown':
      return 'Not checked yet';
    case 'offline':
      return 'Offline';
    case 'warning':
      return h!.failed > 0 ? `${h!.failed} failed` : `${h!.degraded} degraded`;
    default:
      return 'Online';
  }
}

/** Memory in use, as a percentage (undefined when unknown). */
export function memoryPercent(h: Pick<ServerHealth, 'memTotal' | 'memUsed'>): number | undefined {
  return h.memTotal > 0 ? (h.memUsed / h.memTotal) * 100 : undefined;
}

/**
 * Compares two NodeHoster versions ("1.4.2", "v1.10.0-rc1"): negative when
 * a is older. Unknown versions compare equal.
 */
export function compareVersions(a: string | undefined, b: string | undefined): number {
  const parse = (v: string | undefined) => {
    const m = /^v?(\d+)(?:\.(\d+))?(?:\.(\d+))?/.exec((v ?? '').trim());
    return m ? [Number(m[1]), Number(m[2] ?? 0), Number(m[3] ?? 0)] : null;
  };
  const x = parse(a);
  const y = parse(b);
  if (!x || !y) return 0;
  for (let i = 0; i < 3; i++) if (x[i] !== y[i]) return x[i] - y[i];
  return 0;
}

/** A note when a connected server runs another version than this one. */
export function versionNote(remote: string | undefined, local: string | undefined): string | null {
  if (!remote || !local || remote === local) return null;
  const c = compareVersions(remote, local);
  if (c < 0) return `older than this server (${local}): newer features are not available there`;
  if (c > 0) return `newer than this server (${local})`;
  return null;
}

/**
 * The message for an endpoint a connected server does not have (a 404
 * "no such endpoint" from an older version).
 */
export function missingEndpointMessage(name: string, version: string | undefined): string {
  return `${name}${version ? ` runs NodeHoster ${version}, which` : ''} does not have this feature. Update it to use this page there.`;
}

/** Groups a SHA-256 fingerprint by bytes (AB:CD:…) for comparing it with the server's. */
export function formatFingerprint(fp: string | undefined): string {
  if (!fp) return '';
  const hex = fp.replace(/[^0-9a-fA-F]/g, '').toUpperCase();
  return hex.match(/.{1,2}/g)?.join(':') ?? '';
}

/** Normalizes a fingerprint as entered (colons, spaces, "SHA256:"), or null when it is not one. */
export function parseFingerprint(input: string): string | null {
  let s = input.trim();
  if (!s) return '';
  s = s.replace(/^sha-?256\s*:/i, '');
  s = s.replace(/[\s:-]/g, '');
  return /^[0-9a-fA-F]{64}$/.test(s) ? s.toUpperCase() : null;
}

/** Client-side check of a web console URL, as the server normalizes it; null when fine. */
export function serverUrlError(input: string): string | null {
  const raw = input.trim();
  if (!raw) return 'Enter the web console URL, e.g. https://web02:8484';
  let u: URL;
  try {
    u = new URL(raw.includes('://') ? raw : `https://${raw}`);
  } catch {
    return 'Not a URL';
  }
  if (u.username || u.password || u.search || u.hash) return 'The URL cannot have a user name, a query or a fragment';
  if (u.protocol === 'http:') {
    const h = u.hostname.replace(/^\[|\]$/g, '');
    if (h !== 'localhost' && h !== '::1' && !/^127\./.test(h)) return 'Use https://: the API token must not cross the network in clear text';
  } else if (u.protocol !== 'https:') {
    return 'The URL must start with https://';
  }
  return null;
}
